uniffi/src/ffi/rustcalls.rs

```diff
@@ -0,0 +1,218 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+//! # Low-level support for calling rust functions
+//!
+//! This module helps the scaffolding code make calls to rust functions and pass back the result to the FFI bindings code.
+//!
+//! It handles:
+//!    - Catching panics
+//!    - Adapting `Result<>` types into either a return value or an error
+
+use crate::{RustBuffer, ViaFfi};
+use anyhow::Result;
+use ffi_support::IntoFfi;
+use std::panic;
+use std::mem::MaybeUninit;
+
+/// Represents the success/error of a rust call
+///
+/// ## Usage
+///
+/// - The consumer code allocates a `RustCallStatus`
+///    - The error_buf field can either be uninitialized or initialized to zero-capacity and a null
+///      data buffer.  The data here is going to be overwritten so any allocations will leak memory.
+/// - A pointer to this object is passed to the rust FFI function function.  This is an
+///   "out parameter" which we use to write error/success information that occurred during the
+///   function's execution.
+/// - After the call, if code is CALL_ERROR then `error_buf` will be set to a RustBuffer
+///   containing the serialized error object.   The consumer is responsible for freeing
+///   `error_buf`.
+///
+/// ## Layout/fields
+///
+/// The layout of this struct is important since consumers on the other side of the FFI need to
+/// construct it.  If this were a C struct, it would look like:
+///
+/// ```c,no_run
+/// struct RustCallStatus {
+///     int8_t code;
+///     RustBuffer error_buf;
+/// };
+/// ```
+///
+/// #### The `code` field.
+///
+///  - CALL_SUCESS (1) for successful calls
+///  - CALL_ERROR (2) for calls that returned an `Err` value
+///  - CALL_PANIC (3) for calls that panicked
+///
+/// #### The `error_buf` field.
+///
+/// - For CALL_ERROR this is a `RustBuffer` with the serialized error.  The consumer code is
+///   responsible for freeing this `RustBuffer`.
+/// - For other statuses, the value of this field is undefined and the consumer code should ignore
+///   it.
+#[repr(C)]
+pub struct RustCallStatus {
+    // All fields are MaybeUninit for a couple reasons:
+    //
+    //  - When we're passed this struct all fields maybe be uninitialized
+    //  - When we "return" this struct the error_buf field may still be uninitialized
+    pub code: MaybeUninit<i8>,
+    pub error_buf: MaybeUninit<RustBuffer>,
+}
+
+const CALL_SUCCESS: i8 = 1;
+const CALL_ERROR: i8 = 2;
+const CALL_PANIC: i8 = 3;
+
+
+// Generalized rust call handling function
+fn make_call<F, R>(out_status: &mut RustCallStatus, callback: F) -> R::Value
+where
+    F: panic::UnwindSafe + FnOnce() -> Result<R, RustBuffer>,
+    R: IntoFfi,
```

Comment:
I would really love to get rid of the dependency on `ffi_support::InfoFfi` here.  The only thing we're using it for is to define a value that we can return for errors/panics, which the consumer code will just throw away.  I can think of a couple ways, but I'm not sure if they make sense and are in scope for this project:

  - Return a `MaybeUninit<>` value.  This feels like the cleanest solution, but then it would require quite a few changes to the scaffolding templates.
  - Add a `uniffi` trait that defines a error return value for each `ViaFFI::FfiType`.  One issue with this is it would be a breaking change for any consumers that defined their own `ViaFfi` impls with other `FfiType` types.

Reply:
Hrm, interesting. I think you're right the `MaybeUninit` feels like the right thing, but I don't have a good sense of how disruptive that would be.

I suppose an alternative would be to do what Rust does under the hood, and actually return a `union` that can contain the error buffer *or* the success value. But that feels likely to be clunky for common cases where we're returning primitive values.

Reply:
I thought about a union and it could be very nice.  I think the fact that it's going to be sized larger than most primitives should be fine since we get to avoid creating the `RustCallStatus` out param.  The issue I couldn't figure it out was how to deal with all the different return types:
  - You could make a separate union for each return type `union U8Return { value: u8, error: RustBuffer }`, but this is going to be a lot of unions both on the scaffolding and bindings side.
  - You could make a single union that covers all return types, but then if 1 return type is very large then the entire union is going to be inefficient.

One question I had is if we could make any assumptions on which `FfiType` instances can be returned.  I think uniffi only uses integers, `void *`, and `RustBuffer`.  Can we assume that the consumer code isn't going to use those too, or could they return some exotic type?

Reply:
> One question I had is if we could make any assumptions on which FfiType instances can be returned. I think uniffi
> only uses integers, void *, and RustBuffer. Can we assume that the consumer code is going to use those too, or could
> they return some exotic type?

This is probably true in practice right now, but architecturally it feels like we'll make a better longer-term design if we do not assume this.

I also think it would be totally fine to punt this to followup work and continue to use `IntoFfi` for the first round.